vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <map>
+
+#if ! (defined(WIN32) || defined(__APPLE__))
+#include <X11/extensions/XInput2.h>
+#include <X11/extensions/XI2.h>
+#include <X11/XKBlib.h>
+#endif
+
+#include <FL/Fl.H>
+#include <FL/x.H>
+
+#include <rfb/LogWriter.h>
+
+#if !defined(WIN32) && !defined(__APPLE__)
+#include "GestureHandler.h"
+#endif // !defined(WIN32) && !defined(__APPLE__)
+
+#include "touch.h"
+
+static rfb::LogWriter vlog("Touch");
+
+// Sensitivity threshold for gestures
+const int ZOOMSENS = 30;
+const int SCRLSENS = 50;
+
+#if !defined(WIN32) && !defined(__APPLE__)
+static int xi_major;
+
+static bool tracking_touch, grabbed;
+#endif
+
+#if !defined(WIN32) && !defined(__APPLE__)
+int xi2_grabDevices(Window window)
+{
+  int ret, ndevices;
+
+  XIEventMask eventmask;
+  XIDeviceInfo *devices, *device;
+
+  unsigned char flags[XIMaskLen(XI_LASTEVENT)] = { 0 };
+
+  XISetMask(flags, XI_ButtonPress);
+  XISetMask(flags, XI_Motion);
+  XISetMask(flags, XI_ButtonRelease);
+  XISetMask(flags, XI_TouchBegin);
+  XISetMask(flags, XI_TouchUpdate);
+  XISetMask(flags, XI_TouchEnd);
+
+  eventmask.mask = flags;
+  eventmask.mask_len = sizeof(flags);
+
+  devices = XIQueryDevice(fl_display, XIAllMasterDevices, &ndevices);
+
+  for (int i = 0; i < ndevices; i++) {
+    device = &devices[i];
+
+    if (device->use != XIMasterPointer)
+      continue;
+
+    eventmask.deviceid = device->deviceid;
+
+    ret = XIGrabDevice(fl_display,
+                       device->deviceid,
+                       window,
+                       CurrentTime,
+                       None,
+                       XIGrabModeAsync,
+                       XIGrabModeAsync,
+                       True,
+                       &eventmask);
+
+    if (ret) {
+      if (ret == XIAlreadyGrabbed)
+        continue;
+      else {
+        vlog.error("Failure grabbing mouse");
+        return ret;
+      }
+    }
+  }
+
+  XIFreeDeviceInfo(devices);
+
+  grabbed = true;
+
+  return XIGrabSuccess;
+}
+
+void xi2_ungrabDevices()
+{
+  int ndevices;
+  XIDeviceInfo *devices, *device;
+
+  devices = XIQueryDevice(fl_display, XIAllMasterDevices, &ndevices);
+
+  for (int i = 0; i < ndevices; i++) {
+    device = &devices[i];
+
+    if (device->use != XIMasterPointer)
+      continue;
+
+    XIUngrabDevice(fl_display, device->deviceid, CurrentTime);
+  }
+
+  XIFreeDeviceInfo(devices);
+
+  grabbed = false;
+}
+#endif
+
+#if !defined(WIN32) && !defined(__APPLE__)
+class FLTKGestureHandler: public GestureHandler
+{
+public:
+  FLTKGestureHandler(Window wnd);
+
+  void processEvent(const XIDeviceEvent* devev);
+
+protected:
+  //void setTouchEventState(XEvent* dst);
+  void fakeMotionEvent(const XIDeviceEvent* origEvent);
+  void fakeButtonEvent(bool press, int button,
+                       const XIDeviceEvent* origEvent);
+  void prepXEvent(XEvent* dst, const GHEvent src);
+  void prepXEvent(XEvent* dst, const XIDeviceEvent* src);
```

Comment:
These should probably be renamed `preparePointerEvent()` now that we also have keyboard events.

Reply:
bd5163ff